### ISSUE TITLE
SQL Server - can't update views (which I can update using other tools)

### ISSUE THREAD

**Issue by aisbergde** [dbeaver/dbeaver#12131]
#### System information: 
- Operating system (distribution) and version Windows 10
- DBeaver version
- Additional extensions

#### Connection specification:
- Database name and version: SQL Server 2019
- Driver name
- Do you use tunnels or proxies (SSH, SOCKS, etc)? no 

#### Describe the problem you're observing:

Sometimes I use views to update content in tables, for example if I want to add some lookup content.

I can update these views in SSMS or in Azure Data Studio, but I can't update them in dbeaver.

First I am asked to define a virtual PK to identify a unique row.
But even with a virtual key it is not possible to update views. I get this kind of errors: 
![image](https://user-images.githubusercontent.com/9020187/115119482-d1e99380-9fa8-11eb-80d2-945bf332472d.png)

![image](https://user-images.githubusercontent.com/9020187/115119686-ec703c80-9fa9-11eb-91f6-4135f69167a9.png)

The error message is wrong, the object exists. But it looks like dbeaver want to update only tables and thinks the view doesn't exist. But I edited content in this view, it exists.

This is only one example. I was never able to update views.

#### Steps to reproduce, if exist:

code of the tables and of the view:

```sql
CREATE TABLE [repo].[Index_virtual](
	[index_guid] [uniqueidentifier] NOT NULL,
	[has_managedName] [bit] NOT NULL,
	[index_name] [nvarchar](128) NULL,
	[index_type] [tinyint] NOT NULL,
	[is_index_disabled] [bit] NOT NULL,
	[is_index_primary_key] [bit] NOT NULL,
	[is_index_unique] [bit] NOT NULL,
	[parent_RepoObject_guid] [uniqueidentifier] NOT NULL,
	[referenced_index_guid] [uniqueidentifier] NULL,
	[RowNumberInReferencing] [int] NULL,
 CONSTRAINT [PK_Index_virtual] PRIMARY KEY CLUSTERED 
(
	[index_guid] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY]
GO

Create Table [repo].[Index_Settings](
	[index_guid] [UniqueIdentifier] Not Null,
	[IndexPatternColumnDatatype] [NVarchar](4000) Null,
	[IndexPatternColumnName] [NVarchar](4000) Null,
	[IndexSemanticGroup] [NVarchar](512) Null,
	[is_create_constraint] [Bit] Not Null,
 Constraint [PK_Index_Settings] Primary Key Clustered 
(
	[index_guid] Asc
)With (Pad_Index = Off, Statistics_Norecompute = Off, Ignore_Dup_Key = Off, Allow_Row_Locks = On, Allow_Page_Locks = On, Optimize_For_Sequential_Key = Off) On [PRIMARY]
) On [PRIMARY]
Go

Alter Table [repo].[Index_Settings] Add  Default ((0)) For [is_create_constraint]
Go



/*
can be used to update data in repo.Index_virtual
the RepoObject names are shown, can be used for sorting
*/
create View [repo].[Index_virtual_ForUpdate]
As
Select
    iv.index_guid
  , iv.has_managedName
  , iv.index_name
  , iv.index_type
  , iv.is_index_disabled
  , iv.is_index_primary_key
  , iv.is_index_unique
  , iv.parent_RepoObject_guid
  , iv.referenced_index_guid
  , iv.RowNumberInReferencing
  , iset.IndexPatternColumnDatatype
  , iset.IndexPatternColumnName
  , iset.IndexSemanticGroup
  , ro.RepoObject_fullname
  , ro.RepoObject_fullname2
  , ro.SysObject_fullname
  , ro.SysObject_fullname2
  , parent.is_persistence
  , parent.is_repo_managed
From
    repo.Index_virtual        As iv
    Inner Join
        repo.RepoObject       As ro
            On
            ro.RepoObject_guid     = iv.parent_RepoObject_guid

    Left Join
        repo.Index_Settings   iset
            On
            iset.index_guid        = iv.index_guid

    Left Join
        repo.RepoObject_gross As parent
            On
            parent.RepoObject_guid = iv.parent_RepoObject_guid;
Go
```

**Comment by jetzerb** [comment]
@aisbergde : I'm interested in this use case, so decided to look into it for myself, and tried using your repro script above.  Unfortunately, it's incomplete because
- Your view references two tables that you don't provide definitions for
   - RepoObject
   - RepoObject_gross
- You didn't provide any DML statements that generate the error.

But I wanted to take a closer look because it seems strange to me that the results of running an insert/update statement would depend on the tool from which it was run.  So I updated the view definition
- commented out the joins for the missing tables
- and the associated `ro` and `parent` columns in the view definition
 
I then **successfully** issued these statements, which insert and update the `Index_virtual` table:
```sql
insert into repo.index_virtual_forupdate 
(index_guid,has_managedName,index_name,index_type,is_index_disabled,is_index_primary_key,is_index_unique,parent_RepoObject_guid)
values (newid(),0,'',0,0,0,0,newid())
;
update repo.index_virtual_forupdate set index_name = 'abc';
```

Then tried some statements that would affect the joined table as well:
```sql
update repo.index_virtual_forupdate set IndexPatternColumn = 'def'; -- 0 rows affected, as expected

-- insert statement that affects both tables
insert into repo.index_virtual_forupdate 
(index_guid,has_managedName,index_name,index_type,is_index_disabled,is_index_primary_key,is_index_unique,parent_RepoObject_guid,IndexPatternColumnName)
values (newid(),0,'',0,0,0,0,newid(),'ghi')
/* Result is error, as expected, regardless of whether it's run in

DBeaver:
SQL Error [4405] [S0001]: View or function 'repo.index_virtual_forupdate' is not updatable because the modification affects multiple base tables.

or SSMS:
Msg 4405, Level 16, State 1, Line 92
View or function 'repo.index_virtual_forupdate' is not updatable because the modification affects multiple base tables.
*/
```

Can you update your issue with more information?  In my testing, I haven't been able to reproduce your problem.  I'm using version  21.0.3.202104181339.

**Comment by uslss** [comment]
cannot reproduce it against 21.1.0.
If issue is still actual for you - feel free to reopen it with additional info

**Comment by aisbergde** [comment]
@uslss please reopen the issue

Here how to reproduce (MS SQL Server)

## option 1 - restore a database including data from a dacpac file

[dhw_self.zip](https://github.com/dbeaver/dbeaver/files/6816938/dhw_self.zip)

The database is from my open source project https://datahandwerk.github.io/ there is nothing secret.

## create only the required objects to investigate the table

create 2 tables and one view

```sql

CREATE TABLE [repo].[RepoObject](
	[RepoObject_guid] [uniqueidentifier] NOT NULL,
	[has_different_sys_names]  AS (CONVERT([bit],case when [RepoObject_schema_name]<>[SysObject_schema_name] OR [RepoObject_name]<>[SysObject_name] OR [RepoObject_type]<>[SysObject_type] then (1) else (0) end)),
	[has_execution_plan_issue] [bit] NULL,
	[has_get_referenced_issue] [bit] NULL,
	[Inheritance_StringAggSeparatorSql] [nvarchar](4000) NULL,
	[InheritanceDefinition] [nvarchar](4000) NULL,
	[InheritanceType] [tinyint] NULL,
	[is_required_ObjectMerge] [bit] NULL,
	[is_repo_managed] [bit] NULL,
	[is_SysObject_missing] [bit] NULL,
	[modify_dt] [datetime] NOT NULL,
	[pk_index_guid] [uniqueidentifier] NULL,
	[pk_IndexPatternColumnName_new] [nvarchar](4000) NULL,
	[Repo_history_table_guid] [uniqueidentifier] NULL,
	[Repo_temporal_type] [tinyint] NULL,
	[RepoObject_name] [nvarchar](128) NOT NULL,
	[RepoObject_Referencing_Count] [int] NULL,
	[RepoObject_schema_name] [nvarchar](128) NOT NULL,
	[RepoObject_type] [char](2) NOT NULL,
	[SysObject_id] [int] NULL,
	[SysObject_modify_date] [datetime] NULL,
	[SysObject_name] [nvarchar](128) NOT NULL,
	[SysObject_parent_object_id] [int] NOT NULL,
	[SysObject_schema_name] [nvarchar](128) NOT NULL,
	[SysObject_type] [char](2) NULL,
	[is_RepoObject_name_uniqueidentifier]  AS (case when TRY_CAST([RepoObject_name] AS [uniqueidentifier]) IS NULL then (0) else (1) end) PERSISTED NOT NULL,
	[is_SysObject_name_uniqueidentifier]  AS (case when TRY_CAST([SysObject_name] AS [uniqueidentifier]) IS NULL then (0) else (1) end) PERSISTED NOT NULL,
	[node_id]  AS (CONVERT([bigint],[SysObject_id])*(10000)),
	[RepoObject_fullname]  AS (concat('[',[RepoObject_schema_name],'].[',[RepoObject_name],']')) PERSISTED NOT NULL,
	[RepoObject_fullname2]  AS (concat([RepoObject_schema_name],'.',[RepoObject_name])) PERSISTED NOT NULL,
	[SysObject_fullname]  AS (concat('[',[SysObject_schema_name],'].[',[SysObject_name],']')) PERSISTED NOT NULL,
	[SysObject_fullname2]  AS (concat([SysObject_schema_name],'.',[SysObject_name])) PERSISTED NOT NULL,
	[SysObject_query_sql]  AS (concat('SELECT * FROM [',[config].[fs_dwh_database_name](),'].[',[SysObject_schema_name],'].[',[SysObject_name],']')),
	[usp_persistence_name]  AS ('usp_PERSIST_'+[RepoObject_name]) PERSISTED NOT NULL,
 CONSTRAINT [PK_RepoObject] PRIMARY KEY CLUSTERED 
(
	[RepoObject_guid] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY],
 CONSTRAINT [UK_RepoObject__RepoNames] UNIQUE NONCLUSTERED 
(
	[RepoObject_schema_name] ASC,
	[RepoObject_name] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY],
 CONSTRAINT [UK_RepoObject__SysNames] UNIQUE NONCLUSTERED 
(
	[SysObject_schema_name] ASC,
	[SysObject_name] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY]
GO

ALTER TABLE [repo].[RepoObject] ADD  CONSTRAINT [DF_RepoObject_RepoObject_guid]  DEFAULT (newsequentialid()) FOR [RepoObject_guid]
GO

ALTER TABLE [repo].[RepoObject] ADD  CONSTRAINT [DF_RepoObject_modify_dt]  DEFAULT (getdate()) FOR [modify_dt]
GO

ALTER TABLE [repo].[RepoObject] ADD  CONSTRAINT [DF_RepoObject_RepoObject_name]  DEFAULT (newid()) FOR [RepoObject_name]
GO

ALTER TABLE [repo].[RepoObject] ADD  CONSTRAINT [DF_RepoObject_SysObject_name]  DEFAULT (newid()) FOR [SysObject_name]
GO

ALTER TABLE [repo].[RepoObject] ADD  CONSTRAINT [DF_RepoObject_SysObject_parent_object_id]  DEFAULT ((0)) FOR [SysObject_parent_object_id]
GO


CREATE TABLE [property].[RepoObjectProperty](
	[RepoObjectProperty_id] [int] IDENTITY(1,1) NOT NULL,
	[RepoObject_guid] [uniqueidentifier] NOT NULL,
	[property_name] [nvarchar](128) NOT NULL,
	[property_value] [sql_variant] NULL,
	[inheritance] [tinyint] NULL,
	[property_basetype]  AS (sql_variant_property([property_value],'BaseType')),
	[property_int]  AS (case when NOT sql_variant_property([property_value],'BaseType')='uniqueidentifier' then TRY_CAST([property_value] AS [int])  end),
	[property_bigint]  AS (case when NOT sql_variant_property([property_value],'BaseType')='uniqueidentifier' then TRY_CAST([property_value] AS [bigint])  end),
	[property_varchar]  AS (TRY_CAST([property_value] AS [varchar](8000))),
	[property_nvarchar]  AS (TRY_CAST([property_value] AS [nvarchar](4000))),
	[property_real]  AS (case when NOT sql_variant_property([property_value],'BaseType')='uniqueidentifier' then TRY_CAST([property_value] AS [real])  end),
	[property_float]  AS (case when NOT sql_variant_property([property_value],'BaseType')='uniqueidentifier' then TRY_CAST([property_value] AS [float])  end),
	[property_money]  AS (case when NOT sql_variant_property([property_value],'BaseType')='uniqueidentifier' then TRY_CAST([property_value] AS [money])  end),
 CONSTRAINT [PK_RepoObjectProperty] PRIMARY KEY CLUSTERED 
(
	[RepoObjectProperty_id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY],
 CONSTRAINT [UK_RepoObjectProperty] UNIQUE NONCLUSTERED 
(
	[RepoObject_guid] ASC,
	[property_name] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY]
GO

ALTER TABLE [property].[RepoObjectProperty]  WITH CHECK ADD  CONSTRAINT [FK_RepoObjectProperty__RepoObject] FOREIGN KEY([RepoObject_guid])
REFERENCES [repo].[RepoObject] ([RepoObject_guid])
ON UPDATE CASCADE
ON DELETE CASCADE
GO

ALTER TABLE [property].[RepoObjectProperty] CHECK CONSTRAINT [FK_RepoObjectProperty__RepoObject]
GO

create View [property].[RepoObjectProperty_ForUpdate]
As
Select
    T1.[RepoObjectProperty_id]
  , T1.[RepoObject_guid]
  , T1.[property_name]
  , T1.[property_value]
  , T1.[inheritance]
  --,T1.[property_basetype]
  --,T1.[property_int]
  --,T1.[property_bigint]
  --,T1.[property_varchar]
  --,T1.[property_nvarchar]
  --,T1.[property_real]
  --,T1.[property_float]
  --,T1.[property_money]
  , T2.RepoObject_fullname
  , T2.RepoObject_fullname2
From
    [property].[RepoObjectProperty] T1
    Inner Join
        repo.RepoObject             T2
            On
            T2.RepoObject_guid = T1.RepoObject_guid;
Go
```

Fill the table with some data

## reproduce the issue

in dbeaver define a virtual key in the view on [RepoObjectProperty_id]

![image](https://user-images.githubusercontent.com/9020187/125631754-5d7dad6e-5b2e-427e-a05d-692f496fcac8.png)

![image](https://user-images.githubusercontent.com/9020187/125631855-1d7dcd21-1ed9-4866-8b70-2cf49f827dba.png)

try to update the value in  [property].[RepoObjectProperty].[property_value]

When you save, you get 
![image](https://user-images.githubusercontent.com/9020187/125632083-5907f9db-435a-46ef-a5ca-facaa077c0f4.png)

select `Custom unique key`

you get the window from the screenshot above, confirm

and you get this error:

![image](https://user-images.githubusercontent.com/9020187/125632304-8fc60388-c435-4caf-888f-74124c04a745.png)

close the view

reopen the view

now you get some other errors

I change a value from `Objekt_Date` to `Objekt_Date aaa`

And I get this error

![image](https://user-images.githubusercontent.com/9020187/125633022-6ed5027b-1ad8-45fb-8357-28759c2d68f5.png)

![image](https://user-images.githubusercontent.com/9020187/125633227-afc0af30-cfe3-4804-896c-5957c76c2187.png)

**Comment by LonwoLonwo** [comment]
Sorry for the late response.

Thanks for the DDL.
Unfortunately, I can't reproduce the problem with the RepoObjectProperty_ForUpdate update. I created RepoObject and [property].[RepoObjectProperty]. Also, I created a virtual key in the view on [RepoObjectProperty_id]. And added some data in tables.

But. On the step

> try to update the value in [property].[RepoObjectProperty].[property_value]

I have no "choose your key" dialog, because [RepoObjectProperty] has the key.

However, another problem with sql_variant data type binding takes place. I will fix it.
